File: training-data/php/prototype_pattern.php
<?php

abstract class Document
{
    /** @var string[] */
    protected array $tags = [];

    abstract public function clone(): static;

    public function addTag(string $tag): void
    {
        $this->tags[] = $tag;
    }

    public function tags(): array
    {
        return $this->tags;
    }
}

class Report extends Document
{
    public function __construct(public string $title, public array $sections = [])
    {
    }

    public function clone(): static
    {
        $copy = new static($this->title, $this->sections);
        $copy->tags = $this->tags;
        return $copy;
    }
}

$original = new Report('Q3 Summary', ['intro', 'numbers']);
$original->addTag('draft');

$copy = $original->clone();
$copy->title = 'Q3 Summary (copy)';
$copy->addTag('review');

echo "{$original->title}: " . implode(',', $original->tags()) . "\n";
echo "{$copy->title}: " . implode(',', $copy->tags()) . "\n";

File: training-data/go/prototype_pattern.go
package main

import "fmt"

type document struct {
	title    string
	tags     []string
	metadata map[string]string
}

func (d *document) clone() *document {
	tagsCopy := make([]string, len(d.tags))
	copy(tagsCopy, d.tags)

	metaCopy := make(map[string]string, len(d.metadata))
	for k, v := range d.metadata {
		metaCopy[k] = v
	}

	return &document{
		title:    d.title,
		tags:     tagsCopy,
		metadata: metaCopy,
	}
}

func main() {
	original := &document{
		title:    "Report",
		tags:     []string{"draft"},
		metadata: map[string]string{"author": "Alice"},
	}

	copyDoc := original.clone()
	copyDoc.title = "Report Copy"
	copyDoc.tags = append(copyDoc.tags, "duplicate")
	copyDoc.metadata["author"] = "Bob"

	fmt.Println("original:", original.title, original.tags, original.metadata)
	fmt.Println("copy:", copyDoc.title, copyDoc.tags, copyDoc.metadata)
}
